[Merged by Bors] - feat: quantales

PieterCuijpers opened the pull request:
First definition of Quantales - a non-commutative generalization of Locales/Frames. There are still some points of discussion, which would require changes elsewhere in Mathlib, but this definition should be workeable as a starting point (I think).
It's my first PR, but I've tried to stay close to what I've seen in CompleteLattices.

---

<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> Mathlib.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

j-loreaux commented on Mathlib/Order/Quantale.lean:
```diff
@@ -0,0 +1,128 @@
+* class `Quantale`: a monoid distributing over a complete sup-semilattice.
+  I.e satisfying `x * (sSup s) = ⨆ y ∈ s, x * y` and `(sSup s) * y = ⨆ x ∈ s, x * y`
+
+Conceptually, every complete sup-semilattice is of course a complete lattice.
+However, when considering morphisms between quantales, usually only the supremum
+is required to be preserved. Hence, it is conseptually cleaner to define it as a
+complete sup-semilattice. (In literature, both variants occur.)
```
Are you saying that `x * sInf s = ⨅ y ∈ s, x * y` *isn't* always true? Or are you saying that it is, but there are two types of *morphisms* between quantales that are usually considered?

j-loreaux commented on Mathlib/Order/Quantale.lean:
```diff
@@ -0,0 +1,128 @@
+
+/- Basic definitions on quantales -/
+
+/-- A quantale is integral if its unit and top element coincide -/
+def is_integral [Quantale α] := (⊤ : α) = 1
+
+/-- A quantale is commutative if its monoid satisfies `x * y = y * x` -/
+def is_comm [Quantale α] := ∀ x y : α , x * y = y * x
+
+/-- A quantale is idempotent if its monoid satisfies `x * x = x` -/
+def is_idem [Quantale α] := ∀ x : α, x * x = x
```
These should be `IsIntegral`, `IsComm` and `IsIdem`. And presumably you would want them to be `class`es, right? Although, I'm not yet convinced that you want them to be mix-in classes.

For example, you want to conclude that a commutative quantale is a commutative monoid, no? What I'm arguing for here would be `CommQuantale (α : Type*) extends CommMonoid α, CompleteLattice α where ...`

PieterCuijpers replied:
I'll change the spelling, thanks.
Not sure if they need to be classes... what would be the arguments for and against?
And what is a mix-in class?

j-loreaux replied:
Well, if `IsComm` were not a class, then every time you wanted to use the fact about `CommMonoid`s for your commutative quantale, you would need to first activate `Quantale` + `IsComm` → `CommMonoid` instance manually. This would be a pain, and it's probably not what you want. For the others, I'm less sure whether they should be classes, as I don't really know what the use cases are.

Now, as for a mix-in, this is what these classes would be if you just changed them to `class` instead of `def` right now (and, if you did that, you should explicitly mark their type as `Prop`). A mix-in is a class which takes another class as an argument, but only adds `Prop`-valued fields. An example of this in the library would be `IsDomain`. In contrast, we don't use this pattern everywhere. For example, we don't write `Ring R` with `IsComm R` to specify a commutative ring. Instead, we write `CommRing R`, and the structure `CommRing` instead `extends` the structure `Ring`. This means that a `CommRing` carries with it a ring structure, instead of needing one supplied to it.

Reasons for preferring one to the other are related to convenience, performance or parallelism, or yet other reasons. In this case, I would almost certainly argue that `IsComm` should *not* be a mix-in, if for no other reason than we want the parallelism with the rest of the library. You can read more about the algebraic hierarchy in Mathlib in the book "Mathematics in Lean".

PieterCuijpers replied:
My guess is that all properties are "at the same level as Comm" in this case, so I've updated the file to reflect this. Of course, we will only know once we start to actually prove things. I now have based the file mostly on the wikipedia entry on Quantales, which means there is one theorem that "obviously" needs to be proven, namely that every IdemIntegralQuantale is a Order.Frame. Will work on that shortly - but I was hoping the definitions can be approved before that?

j-loreaux commented on Mathlib/Order/Quantale.lean:
```diff
@@ -0,0 +1,128 @@
+variable [Quantale α]
+
+/-- Left- and right- residuation operators on a quantale are similar to the Heyting operator
+    on complete lattices, but for a non-commutative logic.
+    I.e. `x ⇨ₗ y = sSup { z | z * x ≤ y }`.
```
style
```suggestion
on complete lattices, but for a non-commutative logic.
I.e. `x ⇨ₗ y = sSup { z | z * x ≤ y }`.
```

j-loreaux commented on Mathlib/Order/Quantale.lean:
```diff
@@ -0,0 +1,128 @@
+/-- Notation for left-residuation in quantales.
+    I.e. `x ⇨ₗ y = sSup { z | z * x ≤ y }`.
+-/
+infixr:60 " ⇨ₗ " => l_residuation
```
I don't think this should be a global notation. I think it should be `scoped` to the `Quantale` namespace.

PieterCuijpers replied:
Agree. What is the code for that?

j-loreaux replied:
I think it's just `scoped infixr....`, but I might be misremembering.

PieterCuijpers replied:
That seems to do it, indeed. Thnx.

j-loreaux commented on Mathlib/Order/Quantale.lean:
```diff
@@ -0,0 +1,128 @@
+    on complete lattices, but for a non-commutative logic.
+    I.e. `x ⇨ₗ y = sSup { z | z * x ≤ y }`.
+-/
+def l_residuation (x y : α) := sSup { z | z * x ≤ y }
+/-- Notation for left-residuation in quantales.
```
style, we generally leave gaps between declarations, and I would probably just use the word `left` instead of abbreviating it in the name. This name is not too long that it needs abbreviations.
```suggestion
def left_residuation (x y : α) := sSup { z | z * x ≤ y }

/-- Notation for left-residuation in quantales.
```

j-loreaux commented on Mathlib/Order/Quantale.lean:
```diff
@@ -0,0 +1,128 @@
+
+/-- A quantale is a monoid distributing over a complete sup-semilattice.
+
+    Technically, I would prefer to define it as
+    `class Quantale (α : Type*) extends Monoid α, CompleteSemilatticeSup α where ...`
+    however, if I do that currently, the elements ⊤ and ⊥ are not automatically defined.
+    I could do that separately, but it seems more sensible to push the definition
+    of those to CompleteSemmilatticeSup in the library, as they are already making
+    sense there and having them should not break anything (in theory).
+
+    The reason not to consider a quantale as a complete lattice, is because the
+    morphisms between quantales usually only preserve the join and not the inf.
+    In fact, when seen as a generalization of locales, the monoid multiplication
+    usually takes the place of the inf.
```
I think you *really* don't want to go the `CompleteSemilatticeSup` route. Indeed, you will lose out on *many* lemmas about `CompleteLattice`s that way, which will just cause pain. Even if in the theory of quantales one never needs the `inf`, it's still useful to have the full structure around. If a use wants to only provide the `CompleteSemilatticeSup` structure and then have it upgraded to a `CompleteLattice` for free, we have a constructor that does that (which will make the defeqs for `inf` and `sInf` terrible, but as you indicate, it's possible that won't matter).

As for the morphisms, there's no problem: when you get around to defining them, just define them to only preserve the join.

PieterCuijpers replied:
Question is why those lemmas that we would loose have not been formulated on CompleteSemiLattices in the first place. If they really do not hold for CompleteSemiLattices, then it is doubtful whether you would want to use them, since you are doing something conceptually strange (categorically speaking).

PieterCuijpers replied:
I see. So in fact you only have the CompleteSemilatticeSup in order to be able to prove that it is a CompleteLattice, and then build everything on that.

I first thought it was conceptually nicer to `only have a CompleteSemilatticeSup' since my own intuition for using Quantales is that the `inf' is not meaningful. But considering it meaningful or not is, I now realize, a bias on my part - based on how I happen to think of them at this point in time. If I let go of that, and in a library you must be able to support all kinds of uses, then it makes sense to just add everything you can reasonably define, and you end up with complete lattices. Thanks... I learned something today :-)

j-loreaux commented on Mathlib/Order/Quantale.lean:
```diff
@@ -0,0 +1,128 @@
+in a system. Quantales come into play when making an observation may change the system itself.
+Traditionally, one would write `x ⊓ y` to describe making observations `x` and `y` at the same
+time, but when the order of observation becomes important due to changes in the system caused
+by the observation, it makes more sense to write `x * y` or `x + y` and consider `*` or `+` as
+a possibly non-commutative monoid.
```
Unless you use `to_additive`, you're only defining multiplicative quantales below. Maybe you should do both? (Note: you'll need to provide the `class` manually first, and then mark the multiplicative one with `to_additive`. See for example, how `Monoid` and `AddMonoid` are declared.)

PieterCuijpers replied:
Thanks, that is indeed what I want ultimately. But since this is my first PR, I figured I should keep it simple and get used to the process first.

Also, when I was looking at AddMonoid, it was not completely clear to me yet how to_additive works. But that's a next step to figure out for me.

j-loreaux replied:
`to_additive` should mostly "Just Work", except that for `class`es and `structure`s you have to declare them manually separately first (additive one first so that when you add the attribute to the multiplicative one it sees the existing additive declaration). Mostly, it works by finding the right name of the additive lemma and then inserting that everywhere. There is a dictionary for this, which isn't too hard to guess (`one` → `zero`, `mul` → `add`, `pow` → `smul`, etc.). There are edge cases where you need to do something manually, but if you have any trouble, just ask on Zulip.

PieterCuijpers replied:
Ah, apparently I needed to update ToAdditive.Frontend as well, adding `Quantale` as a recognized thing to which to_additive applies. After that, things went smoothly indeed.

j-loreaux reviewed (commented):
Welcome to Mathlib! I've left some comments to get you started. Please let me / us know if you have questions. Also, please see the [commit conventions](https://leanprover-community.github.io/contribute/commit.html) about the PR title and description.

j-loreaux commented:
Yes, generally, although it is also possible to request review. 

I see that you now also consider semi groups, which has led to a proliferation of classes. I think there is yet another possibility to consider here: make the algebraic structure an argument to the class rather than extending it.

So you would write


```lean
class Quantale (\a : Type*) [Semigroup \a] extends CompleteLattice \a where ...`
```

The downside is that you have to write *two* things to get a quantale, but the upside is that you can easily change the underlying algebraic structure without defining all new classes.

So, to get a bare bones quantale, you just write:  but to get a commutative unital one, you simply change this to: `[CommMonoid \a] [Quantale \a]`.

After that, you could potentially make `IsIdem` or `IsIntegral` mixin classes. It depends on how often they come up, and whether they come up in conjunction with each other. This is hard for me to say with my limited knowledge of quantales.

PieterCuijpers commented:
> So you would write
> 
> ```lean
> class Quantale (\a : Type*) [Semigroup \a] extends CompleteLattice \a where ...`
> ```
> 
> The downside is that you have to write _two_ things to get a quantale, but the upside is that you can easily change the underlying algebraic structure without defining all new classes.
> 

That does not seem the best way to go to me. I think the definition I have now for quantales fits best.
As I explain in the preceding comments, the term is not completely consistent within literature, with some authors
using it for a semigroup and others using it for a monoid, so we have to decide on which stance Mathlib will take on it.
But in many papers I actually see it acknowledged that the original is without assuming the unit element, to subsequently
say: but in this paper we use the term to mean a unital quantale. Therefore, from a Mathlib point of view, I guess it is okay
to drag along the OneQuantale whenever a paper you are formalizing says Quantale.

The construct you are proposing would make sense if the general way of looking at it is that "you turn a semigroup into a quantale", but (just like with a ring consisting of two semigroups) that is not the way quantales are being used imo.

Thanks for thinking along though! It really helps to have someone giving feedback :-)

j-loreaux commented:
I think you are getting hung up on interpretations. You don't have to think of it as "turning a monoid into a quantale". Instead, you can think of it as: "a semigroup is the minimal algebraic structure necessary for a quantale".

Just for comparison, `Module R M` takes two type class arguments: `AddCommMonoid M` and `Semiring R`. So, to get a vector space over a field, you have to write `[AddCommGroup V] [Field K] [Module K V]`.

I would encourage you to throw out preconceptions about how things "should" look based on how you think of it before formalization. Sometimes the requirements of formalization make things different.

PieterCuijpers commented:
Thanks... easier said than done though ;-)
It turns out now that the decision to go for additive versions of Quantales causes a clash with Semirings.
The add_idem has been defined there already, and again I have to let go of some preconceptions...

j-loreaux commented:
> In fact, there may be multiple ways to turn a semiring into a sup-semilattice. Indeed, when it is an idempotent semiring, choosing the order to match $$x \leq y$$ iff $x + y = y$$ is a natural option, but the reverse is always an option as well.

> I don't think the formalization of it all forces one direction or another. But I do understand that in some branches of mathematics people have come to think of the structure in one way (as an algebraic property of the +), and in other branches they think of the same structure in another way (as an order-theoretic property of the +). In this case, the latter group of people got to developing that part of the library first, so the first group of people will have to adapt a bit.

I don't agree with this assessment at all. Sure, there are many ways to put a sup-semilattice on a semiring, it's just that the structures won't be related (and you can do that in Lean). But in an idempotent semiring, there's a canonical way to do it so that they *are*. And there's no need to think of one structure generating the other. They both just exist, and the axioms force them to play nice. So there's not necessarily anything here about one group getting more say about how things develop (although that sort of thing surely does happen sometimes).

PieterCuijpers commented:
I've tried out your suggestion of making the definition dependent on the Semigroup.
I like the way it makes the definition more independent.
The problem I encountered with add_idem clashing in Kleene now disappears for the time being, because there is no need to have separate definitions for properties that are actually only targetting the Semigroup. Those should go elsewhere. So only the IsIntegral property stays and is not a mix-in class.

Now I bump into being a novice on Git. I need to figure out how to undo my changes to Kleene and IsIdem to get a clean PR... Any suggestions? (Concretely, I need to restore an old version of the Kleene file and remove the IsIdem file for the PR, preferably in such a way that I can easily put them back afterwards when I continue to actually proof theorems about idempotent quantales - because that is when I need them.)

PieterCuijpers commented:
> > I don't think the formalization of it all forces one direction or another. But I do understand that in some branches of mathematics people have come to think of the structure in one way (as an algebraic property of the +), and in other branches they think of the same structure in another way (as an order-theoretic property of the +). In this case, the latter group of people got to developing that part of the library first, so the first group of people will have to adapt a bit.
> 
> I don't agree with this assessment at all. Sure, there are many ways to put a sup-semilattice on a semiring, it's just that the structures won't be related (and you can do that in Lean). But in an idempotent semiring, there's a canonical way to do it so that they _are_. And there's no need to think of one structure generating the other. They both just exist, and the axioms force them to play nice. So there's not necessarily anything here about one group getting more say about how things develop (although that sort of thing surely does happen sometimes).

I agree. I may not be expressing myself clearly where my concern really is. 

What I ran into during an earlier small project, is that I thought I had defined an arbitrary order structure to work on in combination with an algebraic structure. I had no intention of using the canonical order, which apparently also exists. But only when I got into the proofs I found out Lean interpreted me defining an order structure as well as an algebraic structure as "oh he means the canonical one". So all of a sudden I was provided with axioms that I didn't actually want to have and that invalidated my theorem. Luckily, actually, because the same could happen the other way around. If you don't notice that under the hood there is an order structure defined, you might accidentally use that one while the definitions really seem to suggest something else. Such confusions make me quite uncertain when using Lean, so I try to keep control over what is defined and what is not.

Do you have similar experiences?

YaelDillies commented on Mathlib/Order/Quantale.lean:
```diff
@@ -0,0 +1,159 @@
+/-- An integral (or strictly two-sided) quantale is a quantale over a monoid where
+    top and unit coincide. -/
+@[to_additive]
+class IsIntegral (α : Type*) [Monoid α] [Quantale α] : Prop where
```
This name is way too generic to be placed in the root namespace. Can you rename to something like `IsIntegralQuantale` or `Quantal.IsIntegral`? (slight preference for the first)

Feel free to complain on Zulip that the ring theoretic `IsIntegral` is not namespace, though 😉

PieterCuijpers replied:
Renaming it to Quantale.IsIntegral is not a problem.

Renaming it to IsIntegralQuantale leads to the additive version being IsAddIntegralAddQuantale, which I think is a bit ugly. That can probably be solved in the Frontend of ToAdditive, but I don't know how exactly.

Why do you have a slight preference for IsIntegralQuantale?
(And how about the IsIdem for Semigroups? I guess that should follow the same approach then?)

PieterCuijpers replied:
PS. I can for the time being leave out the Kleene changes, which makes for a smaller PR, but I'm struggling with GIT to do the revert in a nice way. Should I do that, or is it okay to make the update to Kleene in the same PR?

YaelDillies replied:
> Renaming it to IsIntegralQuantale leads to the additive version being IsAddIntegralAddQuantale

Shouldn't the additive version be `IsIntegralAddQuantale`? Note that you can tweak the name that `to_additive` guesses

PieterCuijpers replied:
Got it right now. I updated to_additive earlier, undid that change now, so now your suggestion makes more sense. I've updated it and also applied it to the IsIdemSemigroup.

YaelDillies commented on Mathlib/Algebra/Group/IsIdem.lean:
```diff
@@ -0,0 +1,49 @@
+
+## Main definitions
+
+* class `IsAddIdem`: a additive semigroup satisfying `x + x = x`.
+
+* class `IsIdem`: a semigroup satisfying `x * x = x`.
```
```suggestion
* `IsIdemSemigroup`: Typeclass mixin for a semigroup respecting `x * x = x`.
* `IsAddIdemSemigroup`: Typeclass mixin for a semigroup respecting `x + x = x`.
```

YaelDillies commented on Mathlib/Algebra/Group/IsIdem.lean:
```diff
@@ -0,0 +1,49 @@
+* class `IsAddIdem`: a additive semigroup satisfying `x + x = x`.
+
+* class `IsIdem`: a semigroup satisfying `x * x = x`.
+
+## Naming conventions
+
+## Notation
+
+## References
+
+## TODO
+
+-/
```
```suggestion
-/
```

YaelDillies commented on Mathlib/Algebra/Group/IsIdem.lean:
```diff
@@ -0,0 +1,49 @@
+/-- An idempotent additive semigroup is a type with an associative idempotent `(+)`. -/
+class IsIdemAddSemigroup (G : Type _) [AddSemigroup G] : Prop where
+  /-- Idempotence: `x + x = x` -/
+  protected add_idem : ∀ x : G, x + x = x
```
```suggestion
  protected add_idem (x : G) : x + x = x
```
Same below

YaelDillies commented on Mathlib/Algebra/Group/IsIdem.lean:
```diff
@@ -0,0 +1,49 @@
+
+-/
+
+/-- An idempotent additive semigroup is a type with an associative idempotent `(+)`. -/
```
```suggestion
/-- An idempotent additive semigroup is a type with an associative idempotent addition. -/
```
Same below

YaelDillies commented on Mathlib/Algebra/Group/IsIdem.lean:
```diff
@@ -0,0 +1,49 @@
+variable (G : Type _)
+variable [Semigroup G] [IsIdemSemigroup G]
+
+/-- Making mul_idem available globally -/
```
```suggestion
```
No need to refer to this implementation detail in the docstring

YaelDillies commented on Mathlib/Algebra/Order/Kleene.lean:
```diff
@@ -8,6 +8,7 @@ import Mathlib.Algebra.Ring.InjSurj
 import Mathlib.Algebra.Ring.Pi
 import Mathlib.Algebra.Ring.Prod
 import Mathlib.Tactic.Monotonicity.Attr
+import Mathlib.Algebra.Group.IsIdem
```
Please keep imports sorted

YaelDillies commented on Mathlib/Algebra/Order/Kleene.lean:
```diff
@@ -323,7 +326,7 @@ protected abbrev idemSemiring [IdemSemiring α] [Zero β] [One β] [Add β] [Mul
     add_eq_sup := fun a b ↦ hf <| by erw [sup, add, add_eq_sup]
     bot := ⊥
-    bot_le := fun a ↦ bot.trans_le <| @bot_le _ _ _ <| f a }
+    bot_le := fun a ↦ bot.trans_le <| @bot_le _ _ _ <| f a}
```
```suggestion
    bot_le := fun a ↦ bot.trans_le <| @bot_le _ _ _ <| f a }
```

YaelDillies commented on Mathlib/Order/Quantale.lean:
I would suggest putting this under Algebra.Order

YaelDillies commented on Mathlib/Order/Quantale.lean:
```diff
@@ -0,0 +1,160 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Pieter Cuijpers
+-/
+import Mathlib.Order.CompleteLattice
+import Mathlib.Algebra.Group.Defs
```
```suggestion
import Mathlib.Algebra.Group.Defs
import Mathlib.Order.CompleteLattice
```

YaelDillies commented on Mathlib/Order/Quantale.lean:
```diff
@@ -0,0 +1,160 @@
+
+namespace Quantale
+
+variable {α : Type _}
```
```suggestion
variable {α : Type*}
```
and everywhere else

PieterCuijpers replied:
Happy to comply, but struggling to understand the difference.
Can you explain what is what?

YaelDillies commented on Mathlib/Order/Quantale.lean:
```diff
@@ -0,0 +1,160 @@
+the Heyting operator on complete lattices, but for a non-commutative logic.
+I.e. `x ⇨ₗ y = sSup { z | z + x ≤ y }`.
+"]
+def left_residuation (x y : α) := sSup { z | z * x ≤ y }
```
```suggestion
def leftResiduation (x y : α) := sSup { z | z * x ≤ y }
```
Same below

YaelDillies commented on Mathlib/Order/Quantale.lean:
```diff
@@ -0,0 +1,160 @@
+"]
+def right_residuation (x y : α) := sSup { z | x * z ≤ y }
+
+/-- Notation for left-residuation in quantales.
+    I.e. `x ⇨ₗ y = sSup { z | z * x ≤ y }`.
+-/
+@[to_additive "Notation for left-residuation in quantales.
+    I.e. `x ⇨ₗ y = sSup { z | z + x ≤ y }`. "]
+scoped infixr:60 " ⇨ₗ " => left_residuation
```
This docstring is what people will see when they hover over your notation, so it's not really useful to tell them it's notation. A good shout is to use `inherit_doc` to automagically copy over the docstring of the definition.
```suggestion
@[to_additive (attr := inherit_doc)]
scoped infixr:60 " ⇨ₗ " => left_residuation
```

PieterCuijpers replied:
This does not seem to work entirely. However, the problem seems to be with the use of to_additive on the introduction of notation. When I test the notation on an AddQuantale, it does not seem to have been properly defined while it is for a Quantale. I've updated the code to work properly now by duplicating the notation introduction (once in namespace Quantale, once in namespace AddQuantale) and used the inherit_doc as suggested.

YaelDillies commented:
You still need to merge master. Do you know how to do that?

PieterCuijpers commented:
> You still need to merge master. Do you know how to do that?

Did not realize it was a separate step, next to running the CI, but I found this command: git merge origin master.
It seems to do what you suggest. Had to resolve one merge conflict on the bibliography.
Should be okay now.

YaelDillies commented on Mathlib/Algebra/Order/Quantale.lean:
```diff
@@ -15,7 +15,7 @@ quantum mechanics, and computer science (see e.g. [Vickers1989] and [Mulvey1986]
 The most general definition of quantale occurring in literature, is that a quantale is a semigroup
 distributing over a complete sup-semilattice. In our definition below, we use the fact that
-every complete sup-semlattice is in fact a complete lattice, and make construct defined for
+every complete sup-semlattice is in fact a complete lattice, and make constructs defined for
```
```suggestion
every complete sup-semilattice is in fact a complete lattice, and make constructs defined for
```

YaelDillies commented on Mathlib/Algebra/Group/IsIdem.lean:
```diff
@@ -0,0 +1,38 @@
+import Mathlib.Algebra.Group.Defs
+
+/-!
+# Mixing for Idempotent Semigroups
```
```suggestion
# Idempotent Semigroups
```

YaelDillies commented on Mathlib/Algebra/Group/IsIdem.lean:
```diff
@@ -0,0 +1,38 @@
+-/
+
+/-- An idempotent additive semigroup is a type with an associative idempotent addition. -/
+class IsIdemAddSemigroup (G : Type _) [AddSemigroup G] : Prop where
```
```suggestion
class IsIdemAddSemigroup (G : Type*) [AddSemigroup G] : Prop where
```
same below

YaelDillies commented on Mathlib/Algebra/Group/IsIdem.lean:
```diff
@@ -0,0 +1,38 @@
+
+section IsIdem
+
+variable (G : Type _)
+variable [Semigroup G] [IsIdemSemigroup G]
```
```suggestion
variable {G : Type*} [Semigroup G] [IsIdemSemigroup G]
```

YaelDillies commented on Mathlib/Algebra/Order/Kleene.lean:
```diff
@@ -3,12 +3,14 @@ Copyright (c) 2022 Siddhartha Prasad, Yaël Dillies. All rights reserved.
 import Mathlib.Algebra.Ring.Prod
 import Mathlib.Tactic.Monotonicity.Attr
 
+
```
```suggestion
```

YaelDillies commented on Mathlib/Algebra/Order/Kleene.lean:
```diff
@@ -131,7 +134,8 @@ theorem add_eq_sup (a b : α) : a + b = a ⊔ b :=
 scoped[Computability] attribute [simp] add_eq_sup
 
-theorem add_idem (a : α) : a + a = a := by simp
+/-- As explained in the definition, an idempotent semiring has an idempotent addition -/
```
```suggestion
/-- An idempotent semiring has idempotent addition -/
```

YaelDillies commented on Mathlib/Algebra/Order/Quantale.lean:
```diff
@@ -0,0 +1,160 @@
+variable (α : Type _)
+variable [Semigroup α] [Quantale α]
+
+@[to_additive]
+theorem mul_sSup_eq_iSup_mul : ∀ x : α, ∀ s : Set α, x * sSup s = ⨆ y ∈ s, x * y :=
+  Quantale.mul_sSup_eq_iSup_mul
+
+@[to_additive]
+theorem sSup_mul_eq_iSup_mul : ∀ s : Set α, ∀ y : α, sSup s * y = ⨆ x ∈ s, x * y :=
+  Quantale.sSup_mul_eq_iSup_mul
```
Can we have the `iSup` version?

PieterCuijpers replied:
Not sure which version you are aiming at.
I'm just generalizing the definition on frames here
(see CompleteBooleanAlgebra).

class Order.Frame (α : Type*) extends CompleteLattice α, HeytingAlgebra α where
  /-- `⊓` distributes over `⨆`. -/
  inf_sSup_le_iSup_inf (a : α) (s : Set α) : a ⊓ sSup s ≤ ⨆ b ∈ s, a ⊓ b

Question is, perhaps, whether 'le' is enough or 'eq' is necessary.

PieterCuijpers replied:
As mul is not commutative, we need two of course.

PieterCuijpers replied:
I found the iSup version for Frames and added a number of theorems for Quantales based on what I found for Frames in CompleteBooleanAlgebra.

Now there are iSup and sup versions, as wel as monotonicity (mul_le_mul). Those help to prove the equivalence on residuals you asked for later as well.

YaelDillies commented on Mathlib/Algebra/Order/Quantale.lean:
```diff
@@ -0,0 +1,160 @@
+variable {α : Type _}
+variable [Semigroup α] [Quantale α]
+
+/-- Left- and right- residuation operators on an additive quantale are similar to the Heyting
```
```suggestion
/-- Left- and right-residuation operators on an additive quantale are similar to the Heyting
```

YaelDillies commented on Mathlib/Algebra/Order/Quantale.lean:
```diff
@@ -0,0 +1,160 @@
+-/
+@[to_additive "Left- and right- residuation operators on an additive quantale are similar to
+the Heyting operator on complete lattices, but for a non-commutative logic.
+I.e. `x ⇨ₗ y = sSup { z | z + x ≤ y }`.
```
```suggestion
I.e. `x ⇨ₗ y ≤ z ↔ z + x ≤ y`.
```
and can you add that as a lemma?

PieterCuijpers replied:
Adding it as a lemma is probably a good idea. But I think of the sSup version is the actual definition, and so I would put that in the comment as well. Do you have a different opinion there?

PieterCuijpers replied:
I stand corrected... I've been reading into residuals a bit more and am convinced you are actually right. Will update.

PieterCuijpers replied:
Also, I think it is actually formulated as $z \leq x \Rightarrow\_l y \iff z * x \leq y$.  See also https://www.sciencedirect.com/science/article/pii/S0168007213001553

YaelDillies commented on Mathlib/Algebra/Order/Quantale.lean:
```diff
@@ -0,0 +1,160 @@
+the Heyting operator on complete lattices, but for a non-commutative logic.
+I.e. `x ⇨ₗ y = sSup { z | z + x ≤ y }`.
+"]
+def leftResiduation (x y : α) := sSup { z | z * x ≤ y }
```
```suggestion
def leftResiduation (x y : α) := sSup {z | z * x ≤ y}
```

YaelDillies commented on Mathlib/Algebra/Order/Quantale.lean:
```diff
@@ -0,0 +1,216 @@
+
+## Notation
+
+* `x ⇨ₗ y` : `sSup {z | z * x ≤ y}`, the leftResiduation of `y` over `x`;
```
```suggestion
* `x ⇨ₗ y` : `sSup {z | z * x ≤ y}`, the `leftResiduation` of `y` over `x`;
```

YaelDillies commented on Mathlib/Algebra/Order/Quantale.lean:
```diff
@@ -0,0 +1,216 @@
+  protected sSup_mul_eq_iSup_mul (s : Set α) (y : α) : sSup s * y = ⨆ x ∈ s, x * y
+
+/-- An integral (or strictly two-sided) additive quantale is a quantale over an additive monoid
+    where top and unit coincide. -/
```
We don't indent docstrings
```suggestion
where `⊤` and `1` coincide. -/
```
same below

YaelDillies commented on Mathlib/Algebra/Order/Quantale.lean:
```diff
@@ -0,0 +1,216 @@
+variable [Semigroup α] [Quantale α]
+
+@[to_additive]
+theorem lres_le_iff_mul_le (x y z : α) : x ≤ y ⇨ₗ z ↔ x * y ≤ z := by
```
```suggestion
theorem leftResiduation_le_iff_mul_le (x y z : α) : x ≤ y ⇨ₗ z ↔ x * y ≤ z := by
```
or rename `leftResiduation` to `lres`

YaelDillies commented on Mathlib/Algebra/Order/Quantale.lean:
```diff
@@ -0,0 +1,216 @@
+    exact h1
+
+@[to_additive]
+theorem rres_le_iff_mul_le (x y z : α) : x ≤ y ⇨ᵣ z ↔ y * x ≤ z := by
```
arguments to iffs are usually implicit
```suggestion
theorem rres_le_iff_mul_le {x y z : α} : x ≤ y ⇨ᵣ z ↔ y * x ≤ z := by
```

YaelDillies commented on Mathlib/Algebra/Order/Quantale.lean:
```diff
@@ -0,0 +1,226 @@
+  monotonicity of the multiplication, and basic equivalence theorems for left- and right-
+  residuation.
+
+## Naming conventions
+
```
```suggestion
```

YaelDillies commented on Mathlib/Algebra/Group/IsIdem.lean:
What about you make this file a separate PR, to ease reviewing?

PieterCuijpers replied:
Agreed. I've also separated the definition of IsIntegralQuantale, because to me it is at the same level of IsIdem.
In fact, just like IsIdem only is a mixin on Semigroups, IsIntegral could only be a mixin on Semigroups over a Lattice, or even just a OneTop... since the definition only says `1 = \top`. But that is of later concern.

I made a new branch that is not yet ready for PR, in which I'll develop the notion that Quantales generalize Frames. In that PR, the IsIdem, IsComm, and IsIntegral mixin's would make sense to have.

YaelDillies commented on Mathlib/Algebra/Order/Quantale.lean:
```diff
@@ -0,0 +1,226 @@
+open Function
+
+/-- An additive quantale is an additive semigroup distributing over a complete lattice. -/
+class AddQuantale (α : Type*) [AddSemigroup α] extends CompleteLattice α where
```
Why not make this a Prop-valued mixin?
```suggestion
class IsAddQuantale (α : Type*) [AddSemigroup α] [CompleteLattice α] where
```

YaelDillies commented on Mathlib/Algebra/Order/Quantale.lean:
```diff
@@ -0,0 +1,226 @@
+/-- An integral (or strictly two-sided) quantale is a quantale over a monoid where
+`⊤` and `1` coincide. -/
+@[to_additive]
+class IsIntegralQuantale (α : Type*) [Monoid α] [Quantale α] : Prop where
```
... then this can be
```suggestion
class IsIntegralQuantale (α : Type*) [Monoid α] extends Quantale α where
```

YaelDillies commented on Mathlib/Algebra/Order/Quantale.lean:
```diff
@@ -71,44 +71,45 @@ overhead if a user does not need them.
 `⊤` and `0` coincide. -/
-class IsIntegralAddQuantale (α : Type*) [AddMonoid α] [AddQuantale α] : Prop where
+class IsIntegralAddQuantale (α : Type*) [AddMonoid α] [CompleteLattice α] [IsAddQuantale α] : Prop
+  where
```
```suggestion
class IsIntegralAddQuantale (α : Type*) [AddMonoid α] [CompleteLattice α] extends IsAddQuantale α where
```
since `IsIntegralAddQuantale` belonds to the same hierarchy as `IsAddQuantale`

YaelDillies commented on Mathlib/Algebra/Order/Quantale.lean:
```diff
@@ -0,0 +1,167 @@
+    exact h1
+
+@[to_additive]
+theorem rightMulResiduation_le_iff_mul_le : x ≤ y ⇨ᵣ z ↔ y * x ≤ z := by
+  rw [rightMulResiduation];
```
Do you need this rewrite at all? I feel like this should work
```suggestion
theorem rightMulResiduation_le_iff_mul_le : x ≤ y ⇨ᵣ z ↔ y * x ≤ z where
  mp h1 := by
    apply le_trans (mul_le_mul_left' h1 _)
    simp_all only [mul_sSup_eq_iSup_mul, Set.mem_setOf_eq, iSup_le_iff, implies_true]
  mpr h1 := le_sSup h1
```

PieterCuijpers commented:
> thanks!
> 
> maintainer merge

Thank you!

Is there anything I still need to do now?
And what about future changes? 
After this has been merged, I assume I can simply start a new one?
(For example to implement j-loreaux idea of using a variable alias.)

j-loreaux commented on Mathlib/Algebra/Order/Quantale.lean:
```diff
@@ -0,0 +1,161 @@
+namespace Quantale
+
+/-- An additive quantale is an additive semigroup distributing over a complete lattice. -/
+class IsAddQuantale (α : Type*) [AddSemigroup α] [CompleteLattice α] where
```
I suspect we don't want this definition in the `Quantale` namespace.

PieterCuijpers replied:
Agree.

Took the chance to include your suggestion to use a typeclass alias `Quantale` for `Semigroup CompleteLattice IsQuantale`. I hope that does not disrupt the review process too much.

Was wondering whether the use of `variable?` in a library file is okay, or whether I should unfold that...

j-loreaux commented on Mathlib/Algebra/Order/Quantale.lean:
```diff
@@ -0,0 +1,161 @@
+
+/-- A quantale is a semigroup distributing over a complete lattice. -/
+@[to_additive]
+class IsQuantale (α : Type*) [Semigroup α] [CompleteLattice α] where
```
Likewise

j-loreaux commented on Mathlib/Algebra/Order/Quantale.lean:
```diff
@@ -0,0 +1,161 @@
+"]
+def rightMulResiduation (x y : α) := sSup {z | x * z ≤ y}
+
+@[inherit_doc]
+scoped infixr:60 " ⇨ₗ " => leftMulResiduation
+
+@[inherit_doc]
+scoped infixr:60 " ⇨ᵣ " => rightMulResiduation
+
+@[inherit_doc]
+scoped infixr:60 " ⇨ₗ " => leftAddResiduation
+
+@[inherit_doc]
+scoped infixr:60 " ⇨ᵣ " => rightAddResiduation
```
@YaelDillies how does Lean know which one (additive or multiplicative) to pick here?

PieterCuijpers replied:
I think because I named them Mul and Add respectively.
For clarity, I now updated this and made the namespace explicit.

YaelDillies replied:
Oh, I thought they were in a different namespace, in which case both notation should usually not be available at the same time

PieterCuijpers replied:
It appears that having them in the same namespace was the mistake that actually made it work. I think Lean knows which ones to pick based on type inferrence. When I got them into different namespaces, it started having trouble because to_additive was not able to deduce the namespace.

j-loreaux reviewed (commented):
Thanks @PieterCuijpers !

Just one small change and then it should be ready.

PieterCuijpers commented:
I've decided to remove the `variable_alias` for the time being, awaiting the PR by Julian on VectorSpaces and having the discussion on how to use it there.

This lead to a small change again regarding namespaces in the Quantale file.
I would like the main namespaces to be Quantale and AddQuantale, not IsQuantale and IsAddQuantale.

When I introduced the variable alias for AddQuantale, this also introduced its namespace, so the definition of addLeftResidual could be done through `to_additive`. Now, however, the AddQuantale namespace is not implicitly introduced anymore, which meant I had to pull the introduction of those definitions forward.

Do you agree that using the namespace `Quantale` is better than `IsQuantale`?
Especially with an eye on the future introduction of the variable_alias, I think it is more natural, but at the moment it has a small cost involved in not being able to use `to_additive` for the introduction of leftAddResidual.

YaelDillies commented:
I would personally stick to the `IsQuantale` and `IsAddQuantale` namespaces, with a view towards renaming if ever `Quantale` and `AddQuantale` become a thing

PieterCuijpers commented:
What I think is happening, is that `to_additive` needs to have made the connection between the `Quantale` namespace and the `AddQuantale` namespace, before it can transfer between the two. But there is no way to put the `@[to_additive]` before a namespace declaration, you can only do that in front of a structure or class definition. So only after you've defined a class or structure with the name of the namespace, `to_additive` can do its job.

j-loreaux reviewed (approved):
Thanks! I know this was a long process, but with more experience (and more theorems instead of definitions), it will likely go quicker! Looking forward to having more contributions from you @PieterCuijpers 

bors merge
